### PR TITLE
[core] Improve Parser interface

### LINKED ISSUE #2602: [core] Remove ParserOptions
Parser options for now are just implemented for decoration. There's no API for PMD to configure them, and they can only be used if you use the parser directly. 
* It will likely be never supported to have a rule choose special parser options. This is because, if several rules need different sets of options, we'd have to parse files several time. 
* There would also be several ways to parse trees, meaning it's harder to get a hand on the API and start writing rules.
* Nontrivial ParserOptions are implemented in the following modules only:
  * Ecmascript:
https://github.com/pmd/pmd/blob/11c2b201c1ffbb2079ade7c3cbfa356c44b6131d/pmd-javascript/src/main/java/net/sourceforge/pmd/lang/ecmascript/EcmascriptParserOptions.java#L60-L68
    * The rhino version should be translated to a LanguageVersion, not a ParserOption
    * Preserving comments is done by default and we could just remove those options
  * [x] [[javascript] Remove JS parser options #2837](https://github.com/pmd/pmd/pull/2837)
  * XML:
https://github.com/pmd/pmd/blob/11c2b201c1ffbb2079ade7c3cbfa356c44b6131d/pmd-xml/src/main/java/net/sourceforge/pmd/lang/xml/XmlParserOptions.java#L23-L43
    * The ASTs PMD produces should be "low-level" enough that code style issues can be detected. Coalescing CDATA nodes, or expanding entity references, defeats this purpose.
    *  PMD should probably not be attempting to validate schemas *in the parser*. If anything schema validation should maybe be a rule.
    * Namespace awareness is kind of like symbol resolution in other languages and should be kept enabled by default
  * [x] #2971 [xml] Remove XML parser options

I think parser options can go away completely, and language properties (#2518) may be used to replace its only functionality (the comment marker). This would also allow suppress markers to be set per-language, as a happy consequence.

On master we need to
- [x] Deprecate ParserOptions and its language-specific subclasses (Apex, Ecmascript, XML)
- [x] Deprecate Rule::getParserOptions
Both done with #2675 

It's still ok to use ParserOptions on master (though not the language-specific subclasses), because LanguageVersionHandler::getParser takes a ParserOptions parameter for the suppress marker. In 7.0 we should however introduce an overload that takes no parameter.

### PR DESCRIPTION
**Pull request by oowekyala**
## Describe the PR

This is the next step for pmd 7.

* Replace the parameters Parser#parse with a ParserTask object. This contains all the info that is passed to the parser, and may be evolved in the future. Parser is now a functional interface.
* ParserOptions are removed. Their only remaining use, as a container for properties, is shifted onto a PropertySource that the ParserTask carries. This is transitional as they'll be converted to language properties
* The Parser interface is moved to the `nspmd.lang.ast` package instead of just `lang`, AbstractParser is removed as it's now useless
* Nodes have access to their language version and the name of the file. These new methods are meant to replace the statefulness of RuleContext, which we can remove in a future PR
  * This is meant to be replaced by TextDocuments in a future PR. The info is contained in an AstInfo object that is set on the root node, that way, it can also be evolved with fewer code changes. 
* Move pmd-java's SemanticChecksLogger into pmd-core. ParserTasks contain one such object, the parser may use it to validate the produced trees. This is not entirely implemented, we use a noop for now.
* Introduce a supertype for PMD's exceptions: FileAnalysisException. These are the exceptions that may be thrown by a SourceCodeProcessor, eg ParseException or TokenMgrError. 
* Now SourceCodeProcessor reads the whole file into a string before starting the parser (the string is passed through the ParserTask). That way, the implementation of Parser does not need to deal with IOExceptions.

So, most of this has holes that need to be filled-in later, but I think it's necessary groundwork.

## Related issues

- Fixes #2602 

## Ready?

- [ ] Added unit tests for fixed bug/feature
- [ ] Passing all unit tests
- [ ] Complete build `./mvnw clean verify` passes (checked automatically by travis)
- [ ] Added (in-code) documentation (if needed)

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to pmd/7.0.x:<br />
This changeset changes 0 violations,<br />
introduces 0 new violations, 0 new errors and 0 new configuration errors,<br />
removes 0 violations, 1 errors and 0 configuration errors.<br />
<a href="https://chunk.io/pmd/98212a95d16045bb96b9e9dc843ffb04/diff1/index.html">Full report</a></td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to master:<br />
This changeset changes 1606 violations,<br />
introduces 4187 new violations, 1 new errors and 0 new configuration errors,<br />
removes 6137 violations, 10 errors and 2 configuration errors.<br />
<a href="https://chunk.io/pmd/98212a95d16045bb96b9e9dc843ffb04/diff2/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by oowekyala** [comment]
@adangel I think I'm going to merge this pr this weekend, and then merge the 7.0.x branch back into #2830

File: pmd-core/src/main/java/net/sourceforge/pmd/lang/Parser.java [hunk 1]
```diff
@@ -1,49 +0,0 @@
...
- * be stateless (which makes them trivially threadsafe).
- *
- * TODO
- *  - Ideally ParserOptions would be an argument to ::parse
- *  - ::parse would also take some more parameters, eg an error collector
- *  - The reader + filename would be a TextDocument
- *
- * @author Pieter_Van_Raemdonck - Application Engineers NV/SA - www.ae.be
- */
-public interface Parser {
```
Comment (adangel) [inline pmd-core/src/main/java/net/sourceforge/pmd/lang/Parser.java:23]: We could keep a interface at the old package to make it easier to transition:

```java
package net.sourceforge.pmd.lang;

@Deprecated
@DeprecatedUntil700
public interface Parser extends net.sourceforge.pmd.lang.ast.Parser {
}
```

Not sure, if that's worth. Regarding these `@DeprecatedUntil700` - we could move these files into a own module called "pmd-compat6".

File: pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/TokenMgrError.java [hunk 2]
```diff
@@ -12,26 +12,27 @@
 /**
  * An error thrown during lexical analysis of a file.
  */
-public final class TokenMgrError extends RuntimeException {
+public final class TokenMgrError extends FileAnalysisException {
```
Comment (adangel) [inline pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/TokenMgrError.java:15]: At some point, we should probably rename that from Error to Exception...

File: pmd-swift/src/main/java/net/sourceforge/pmd/lang/swift/ast/SwiftRootNode.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+import org.antlr.v4.runtime.ParserRuleContext;
+
+import net.sourceforge.pmd.lang.ast.AstInfo;
+import net.sourceforge.pmd.lang.ast.Parser.ParserTask;
+import net.sourceforge.pmd.lang.ast.RootNode;
+import net.sourceforge.pmd.lang.swift.ast.SwiftParser.SwTopLevel;
+
+// package private base class
+abstract class SwiftRootNode extends SwiftInnerNode implements RootNode {
```
Comment (adangel) [inline pmd-swift/src/main/java/net/sourceforge/pmd/lang/swift/ast/SwiftRootNode.java:15]: I was wondering, how/when this new class is actually used. But figured out -> antlr4-wrapper.xml

**Pull request merged**